File: 03_MPI_OpenMP/14_ArraySum/ArraySum.c
#include <stdio.h>
#include <omp.h>

int main() {

    int n, i, sum = 0;

    printf("Enter number of elements: ");

    scanf("%d", &n);

    int arr[n];

    printf("Enter array elements:\n");

    for(i = 0; i < n; i++) {

        scanf("%d", &arr[i]);
    }

    #pragma omp parallel
    {
        int local_sum = 0;

        int thread_id = omp_get_thread_num();

        int total_threads = omp_get_num_threads();

        int start = thread_id * n / total_threads;

        int end = (thread_id + 1) * n / total_threads;

        for(i = start; i < end; i++) {

            local_sum += arr[i];
        }

        printf("Thread %d calculated partial sum = %d\n",
                thread_id, local_sum);

        #pragma omp critical
        {
            sum += local_sum;
        }
    }

    printf("\nFinal Sum = %d\n", sum);

    return 0;
}

File: 03_MPI_OpenMP/14_ArraySum/ArraySum.java
import java.util.Scanner;

public class ArraySum extends Thread {

    static int arr[];
    static int sum = 0;

    int start, end;
    int localSum = 0;

    ArraySum(int start, int end) {

        this.start = start;
        this.end = end;
    }

    public void run() {

        for(int i = start; i < end; i++) {

            localSum += arr[i];
        }

        System.out.println(Thread.currentThread().getName() +
                " calculated partial sum = " + localSum);

        synchronized(ArraySum.class) {

            sum += localSum;
        }
    }

    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");

        int n = sc.nextInt();

        arr = new int[n];

        System.out.println("Enter array elements:");

        for(int i = 0; i < n; i++) {

            arr[i] = sc.nextInt();
        }

        int numThreads = 4;

        ArraySum threads[] = new ArraySum[numThreads];

        int part = n / numThreads;

        for(int i = 0; i < numThreads; i++) {

            int start = i * part;

            int end;

            if(i == numThreads - 1) {

                end = n;
            }
            else {

                end = start + part;
            }

            threads[i] = new ArraySum(start, end);

            threads[i].start();
        }

        for(int i = 0; i < numThreads; i++) {

            threads[i].join();
        }

        System.out.println("\nFinal Sum = " + sum);

        sc.close();
    }
}